### PR TITLE
Add MacOS support for maven build GitHub action

### LINKED ISSUE #94: Add GitHub action for maven build in Mac OS
Add a github action to run maven build and tests on Mac OS when:
 - A PR is opened against master
 - When a commit is pushed to master

### LINKED ISSUE #95: Add ShellUtility support for Mac OS and ZSH
Add ability to support Mac OS and ZSH as well.

### PR DESCRIPTION
**Pull request by the-c0d3br34k3r**
```
Added Mac OS X support for ShellUtility and ShellUtilityTest

Closes #95.
Added ZSH support to ShellUtility so that it can be supported in
MAC OS X.
```
```
Skip WatchService related tests for Mac OS X

Closes #94.
Currently there's a known OpenJDK bug for WatchService (https://bugs.openjdk.java.net/browse/JDK-7133447)
where it doesn't or takes a very long time to generate filesystem events.
Since the WatchService API for Mac OS X isn't reliable, we are skipping
testing it and will not support this.
```
```
Added a GitHub action to run a maven build against Mac OS X

Closes #94.
Added a github action to run maven build and tests on Mac OS when:
 - A PR is opened against master
 - When a commit is pushed to master
```

### CODE REVIEW

File: src/main/java/utilities/OS.java [hunk 1]
```diff
@@ -1,21 +1,38 @@
...
+        OS detectedOS;
+        if(osString.contains("LINUX")) {
+            detectedOS = LINUX;
+        }
+        else if(osString.contains("WINDOWS")) {
+            detectedOS = WINDOWS;
         }
-        else if(os.contains("WINDOWS")) {
-            return WINDOWS;
+        else if(osString.contains("MAC OS X")) {
```
Comment (abprash) [inline src/main/java/utilities/OS.java:29]: Curious, will this work on all mac versions ? Online, i see, they check presence only for "MAC" in it

File: src/main/java/utilities/ShellUtility.java [hunk 2]
```diff
@@ -325,21 +366,32 @@ public static Response executeCommand(Command command, TypeOfShell typeOfShell)
      * @throws TimeoutException     Thrown f the command execution exceeds specified timeout
```
Comment (abprash) [inline src/main/java/utilities/ShellUtility.java:366]: thrown if

File: src/main/java/utilities/OS.java [hunk 3]
```diff
@@ -1,21 +1,38 @@
...
-        if(os.contains("LINUX")) {
-            return LINUX;
+        String osString = System.getProperty("os.name")
+                .toUpperCase();
+        logger.debug("OS string: " + osString);
+        OS detectedOS;
+        if(osString.contains("LINUX")) {
+            detectedOS = LINUX;
+        }
+        else if(osString.contains("WINDOWS")) {
```
Comment (abprash) [inline src/main/java/utilities/OS.java:26]: let's have a linter to take care of these formatting later.

**Comment by abprash** [comment]
Better to split it into atomic PRs

File: src/main/java/utilities/ShellUtility.java (hunk 2 above)
Comment (the-c0d3br34k3r) [reply src/main/java/utilities/ShellUtility.java:365]: Done.

File: src/main/java/utilities/OS.java (hunk 3 above)
Comment (the-c0d3br34k3r) [reply src/main/java/utilities/OS.java:26]: Agreed.

File: src/main/java/utilities/OS.java (hunk 1 above)
Comment (the-c0d3br34k3r) [reply src/main/java/utilities/OS.java:29]: I tested it on the GitHubAction runner, it works. We can fix it later if poses an issue to certain Mac versions.

**Comment by the-c0d3br34k3r** [comment]
> Better to split it into atomic PRs

As per our conversation for the previous PR, we agreed on having relevant changes in the same PR.

File: src/main/java/utilities/ShellUtility.java [hunk 4]
```diff
@@ -349,28 +358,39 @@ public static Response executeCommand(Command command, TypeOfShell typeOfShell)
      * Executes the specified command
      *
      * @param command         Command to execute
-     * @param timeOutDuration Time out for the cmmand execution
+     * @param timeOutDuration Time out for the command execution
```
Comment (silvs110) [inline src/main/java/utilities/ShellUtility.java:361]: Do we need to specify if the timeout is in seconds/milliseconds?

File: src/test/java/FileUtilityTest.java [hunk 5]
```diff
@@ -1,31 +1,49 @@
...
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.CsvSource;
 import utilities.FileUtility;
 import utilities.I18NUtility;
+import utilities.OS;
 import utilities.PropertyUtility;
 import utilities.ShellUtility;
+
```
Comment (silvs110) [inline src/test/java/FileUtilityTest.java:14]:
```suggestion

```

File: src/test/java/FileUtilityTest.java [hunk 6]
```diff
@@ -1,31 +1,49 @@
...
 import java.util.concurrent.TimeoutException;
 import java.util.function.Function;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Set;
+import java.util.TreeSet;
+
```
Comment (silvs110) [inline src/test/java/FileUtilityTest.java:38]:
```suggestion

```

File: src/main/java/utilities/OS.java [hunk 7]
```diff
@@ -1,21 +1,38 @@
...
 public enum OS {
     WINDOWS,
     LINUX,
+    MAC_OS_X,
     UNKNOWN;
 
+    private static Logger logger = LogManager.getLogger(OS.class);
+
+    /**
+     * Detects the OS of system in which this method is called.
```
Comment (silvs110) [inline src/main/java/utilities/OS.java:15]: 'system' is redundant

**Review by silvs110 — CHANGES_REQUESTED** [review]

File: src/main/java/utilities/ShellUtility.java (hunk 4 above)
Comment (the-c0d3br34k3r) [reply src/main/java/utilities/ShellUtility.java:361]: Not needed as the unit will be specified inside the Duration object.

File: src/test/java/FileUtilityTest.java (hunk 5 above)
Comment (the-c0d3br34k3r) [reply src/test/java/FileUtilityTest.java:14]: Will be addressed in a future task for linting.

File: src/test/java/FileUtilityTest.java (hunk 6 above)
Comment (the-c0d3br34k3r) [reply src/test/java/FileUtilityTest.java:38]: Will be addressed in a future task for linting.

**Review by silvs110 — APPROVED** [review]

**Pull request merged**